csrc/preseg_passes/mark_aliases_prepare.cpp

```diff
@@ -50,86 +101,101 @@ void MarkAliasesPreparePass::runPass(Fusion* fusion) {
-  // This algorithm is suboptimal in many cases. See
-  // https://github.com/NVIDIA/Fuser/issues/2395#issuecomment-2207043749 for a
-  // proposal that may work better.
-  std::unordered_set<TensorView*> aliased_ios;
-  for (TensorView* out_tv :
-       ir_utils::filterByType<TensorView>(fusion->outputs())) {
-    if (TensorView* aliased_io = analysis.getNearestAliasedIo(out_tv)) {
-      if (analysis.getNearestAliasedIo(aliased_io) == nullptr) {
-        aliased_ios.insert(aliased_io);
-      }
-    } else {
-      non_alias_outs.push_back(out_tv);
+  // An important detail: a meta op preceding a non-meta op (i.e. has a
+  // non-alias output TensorView) is treated as non-meta for the purpose of
+  // bookend. This is to avoid over segmentation. For example, in
+  //
+  //   N/M -> M1 -> N/M
+  //          |
+  //          -> M2
+  //
+  // we want to avoid putting a `segment_set` before M1, a meta op, because
+  // that would lead to two kernels. See AliasTest.DoNotOverSegment_* for more
+  // examples.
+  const std::unordered_set<Expr*>& used_by_non_aliases =
+      exprsUsedByNonAliases(analysis, fusion);
+  std::vector<Use> uses_to_segment;
+  uses_to_segment.reserve(fusion->outputs().size());
+  for (auto* out : ir_utils::filterByType<TensorView>(fusion->outputs())) {
+    Use use_to_segment = findUseToSegment(out, analysis, used_by_non_aliases);
+    if (use_to_segment.use_of != out) {
+      uses_to_segment.push_back(use_to_segment);
     }
   }
 
-  // Mark all expressions that are (transitively) used by non-alias outputs.
-  auto used_by_non_aliases =
-      [](const std::vector<Expr*>& exprs) -> std::unordered_set<Expr*> {
-    return {exprs.begin(), exprs.end()};
-  }(StmtSort::getExprsTo(non_alias_outs));
-
-  for (TensorView* aliased_io : aliased_ios) {
-    // Divide the users of aliased_io into two groups according to
-    // `used_by_non_aliases`.
-    std::vector<Expr*> users_used_by_non_aliases;
-    std::vector<Expr*> users_used_only_by_aliases;
-    for (Expr* e : aliased_io->uses()) {
-      if (used_by_non_aliases.count(e)) {
-        users_used_by_non_aliases.push_back(e);
-      } else {
-        users_used_only_by_aliases.push_back(e);
-      }
+  // The remaining are optimizations to reduce the number of `segment_set`s
+  // inserted.
+  //
+  // Group `uses_to_segment` by `use_of` and remove duplicates.
+  std::sort(uses_to_segment.begin(), uses_to_segment.end());
+  uses_to_segment.erase(
+      std::unique(uses_to_segment.begin(), uses_to_segment.end()),
+      uses_to_segment.end());
+
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    for (const auto& use : uses_to_segment) {
+      debug() << "Will put a segment_set at " << use.user << std::endl;
     }
+  }
 
-    // If all users are used by non-alias outputs, do nothing. Adding
-    // segment_set around unlikely creates more no-op regions.
-    if (users_used_only_by_aliases.empty()) {
-      continue;
+  auto i = uses_to_segment.begin();
```

Comment:
can we change `i` and `j` to more meaningful names? and add comments to these two while loops.